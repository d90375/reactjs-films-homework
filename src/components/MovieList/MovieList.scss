.main {
  background-color: #f5f6f6;
  padding-top: 4rem;
  padding-left: 4rem;
  padding-right: 4rem;
}

.card__wrap {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
}

.noResults__wrap {
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.noResults__text {
  text-transform: capitalize;
  font-family: "Helvetica Neue", serif;
  font-size: 1rem;
  font-weight: 600;
  line-height: 44px;
  letter-spacing: 1.12px;
  color: #0aaee4;
  margin-bottom: 10rem;
}
