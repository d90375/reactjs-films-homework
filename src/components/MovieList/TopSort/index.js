import TopSort from "./TopSort.container";

export default TopSort;
