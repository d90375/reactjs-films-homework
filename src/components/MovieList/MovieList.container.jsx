import React, { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";

import TopSort from "./TopSort";
import Preloader from "./Preloader";
import MovieList from "./MovieList";
import VideoFrame from "../VideoFrame";

import {
  fetchGenresData,
  getMovieSelector,
  fetchMovieDataFilter,
  getCompletedMovieListSelector
} from "../../modules/movieListData";
import { NUMBER_OF_CARDS } from "../../constants";

const MovieListContainer = () => {
  const dispatch = useDispatch();
  const data = useSelector(getCompletedMovieListSelector);
  const { isLoadingMovieList, hasErrorMovieList, isFulfilledMovieList, errorMovieList } = useSelector(getMovieSelector);

  const [isDisplayCardDirection, setDisplayCardDirection] = useState("square");

  if (data?.results?.length > NUMBER_OF_CARDS) {
    data.results.length = NUMBER_OF_CARDS;
  }

  // Load static genres of the list:
  useEffect(() => {
    dispatch(fetchMovieDataFilter());
    dispatch(fetchGenresData());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const handleSwitchToSquare = () => {
    setDisplayCardDirection("square");
  };
  const handleSwitchToLine = () => {
    setDisplayCardDirection("line");
  };

  return (
    <>
      <VideoFrame />
      <TopSort
        handleSwitchToSquare={handleSwitchToSquare}
        handleSwitchToLine={handleSwitchToLine}
        isDisplayCardDirection={isDisplayCardDirection}
        genres={data?.genres}
      />
      {isLoadingMovieList && <Preloader />}
      {isFulfilledMovieList && (
        <MovieList
          isDisplayCardDirection={isDisplayCardDirection}
          data={data?.results}
          cardLength={data?.results?.length}
          genres={data?.genres}
        />
      )}
      {hasErrorMovieList && (
        <div className="error error__movieList">
          <span>Error {errorMovieList}</span>
        </div>
      )}
    </>
  );
};

export default MovieListContainer;
