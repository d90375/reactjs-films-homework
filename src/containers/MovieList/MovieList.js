import React from "react";
import { useSelector } from "react-redux";
import TopSort from "../../components/Main/TopSort";
import Preloader from "../../components/Main/Preloader";
import Main from "../../components/Main";
import { getDataReselctor } from "../../modules/data/dataSelectors";

const MovieList = () => {
  const data = useSelector(getDataReselctor);
  return (
    <>
      <TopSort />
      <Main data={data} />
      <Preloader />
    </>
  );
};

export default MovieList;
