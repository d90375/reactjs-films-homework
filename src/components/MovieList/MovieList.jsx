import PropTypes from "prop-types";
import React from "react";
import Card from "./Card";
import "./MovieList.scss";

const MovieList = ({ data, cardLength, genres, isDisplayCardDirection }) => {
  return (
    <div className="movieList">
      {cardLength > 0 ? (
        Object.values(data).map((card, cardIndex) => (
          <Card
            key={card.id}
            cardItem={card}
            genresItems={genres}
            cardIndex={cardIndex}
            isDisplayCardDirection={isDisplayCardDirection}
          />
        ))
      ) : (
        <div className="noResults__wrap">
          <span className="noResults__text">No results</span>
        </div>
      )}
    </div>
  );
};

export default MovieList;

MovieList.propTypes = {
  cardLength: PropTypes.number,
  data: PropTypes.arrayOf(
    PropTypes.shape({
      adult: PropTypes.bool,
      backdrop_path: PropTypes.string,
      genre_ids: PropTypes.arrayOf(PropTypes.number),
      id: PropTypes.number,
      original_language: PropTypes.string,
      original_title: PropTypes.string,
      overview: PropTypes.string,
      popularity: PropTypes.number,
      poster_path: PropTypes.string,
      release_date: PropTypes.string,
      title: PropTypes.string,
      video: PropTypes.bool,
      vote_average: PropTypes.number,
      vote_count: PropTypes.number
    })
  ),
  genres: PropTypes.objectOf(PropTypes.string)
};

MovieList.defaultProps = {
  cardLength: 0,
  data: [],
  genres: {}
};
