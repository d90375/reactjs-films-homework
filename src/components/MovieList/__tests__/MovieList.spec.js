import React from "react";
import MovieList from "../MovieList";

describe("Main", () => {
  let component;

  const mockProps = {
    data: [{ data: "data", id: 1 }],
    genres: { 12: "12" }
  };

  it("should render Main component without props", () => {
    component = shallow(<MovieList />);
    expect(component).toMatchSnapshot();
  });

  it("should render Main component without cards ", () => {
    component = shallow(<MovieList cardLength={0} {...mockProps} />);
    expect(component).toMatchSnapshot();
  });

  it("should render Main component with cards", () => {
    component = shallow(<MovieList cardLength={5} {...mockProps} />);
    expect(component).toMatchSnapshot();
  });
});
