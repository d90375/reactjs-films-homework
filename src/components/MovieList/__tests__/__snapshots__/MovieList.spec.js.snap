// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should render Main component with cards 1`] = `
<div
  className="movieList"
>
  <Memo()
    cardIndex={0}
    cardItem={
      Object {
        "data": "data",
        "id": 1,
      }
    }
    genresItems={
      Object {
        "12": "12",
      }
    }
    isDisplayCardDirection="line"
    key="1"
  />
</div>
`;

exports[`Main should render Main component without cards  1`] = `
<div
  className="movieList"
>
  <div
    className="noResults__wrap"
  >
    <span
      className="noResults__text"
    >
      No results
    </span>
  </div>
</div>
`;

exports[`Main should render Main component without props 1`] = `
<div
  className="movieList"
>
  <div
    className="noResults__wrap"
  >
    <span
      className="noResults__text"
    >
      No results
    </span>
  </div>
</div>
`;
